System defined taxonomies

ChrisChV opened the pull request:
## Description

Subclasses related to build System-defined taxonomies:

- [Docs updated](https://github.com/openedx/openedx-learning/pull/67/files#diff-e20a5299480027adca1a87aac4dea6d00797403cc0a23aab43bf0204ddb3a069) to reflect the new architecture of taxonomies.
- `models.py` moved to `models/base.py`
- `tag_ref` [setter](https://github.com/openedx/openedx-learning/pull/67/files#diff-7ffb9100f9426c81c59e88e9beceb9e3a8089da152d1d605b7de97990e6f0d50R540) to allow override how to set a tag on an object_tag
- `object_tag` [added to Taxonomy](https://github.com/openedx/openedx-learning/pull/67/files#diff-7ffb9100f9426c81c59e88e9beceb9e3a8089da152d1d605b7de97990e6f0d50R183)
- Base system-defined subclasses added.
- Language tags fixture.
- Taxonomy.system_defined updated to be a [hardcoded property](https://github.com/openedx/openedx-learning/pull/67/commits/1847248b48176014b989a25e03ec00b9dda5f26e#diff-7ffb9100f9426c81c59e88e9beceb9e3a8089da152d1d605b7de97990e6f0d50R203).
- TaxonomyAdmin updated to avoid edit and delete system taxonomies: https://github.com/openedx/openedx-learning/pull/67/commits/1847248b48176014b989a25e03ec00b9dda5f26e
- TagAdmin updated to avoid edit and delete tags from system taxonomies: https://github.com/openedx/openedx-learning/pull/67/commits/1847248b48176014b989a25e03ec00b9dda5f26e
- TagAmdin updated to avoid create Tags on system taxonomies: https://github.com/openedx/openedx-learning/pull/67/commits/1847248b48176014b989a25e03ec00b9dda5f26e

## Testing instructions

Manual testing is possible but not very useful, since we don't yet have views for these models and APIs.

But you can check that the added tests cover the expected polymorphism behavior supported by this change, and that tests are still covering 100% of the openedx_tagging module.

### Testing admin:

- On the backend shell create Taxonomies an Tags:

```
from openedx_tagging.core.tagging.models import Taxonomy, Tag
closed_taxonomy = Taxonomy(name='Test1', allow_free_text=False)
closed_taxonomy.save()
open_taxonomy = Taxonomy(name='Test2', allow_free_text=True)
open_taxonomy.save()
system_taxonomy = Taxonomy(name='Language', allow_free_text=False)
from openedx_tagging.core.tagging.models import LanguageTaxonomy
system_taxonomy.taxonomy_class = LanguageTaxonomy
system_taxonomy.save()
tag = Tag(taxonomy=closed_taxonomy, value='test')
tag.save()
tag = Tag(taxonomy=system_taxonomy, value='es')
tag.save()
```

- Go to [Django admin](http://127.0.0.1:8000/admin/oel_tagging/taxonomy/) and check that you can see all taxonomies.
- Click on the language taxonomy and check that you cannot edit or delete the taxonomy.
- Go to [Tags](http://127.0.0.1:8000/admin/oel_tagging/tag/) and check that you can see all tags.
- Click on the language tag and check that you cannot edit or delete the taxonomy.
- Create a Tag and check that you cannot add the tag to the open taxonomy nor system taxonomy

## Other information

Relates to https://github.com/openedx/modular-learning/issues/77

ChrisChV committed:
feat: System-defined taxonomies

ChrisChV committed:
feat: Language and Author taxonomies

ChrisChV committed:
feat: Generic system defined object tags and Language object tag added

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,172 @@
+        Returns a list of tags of the available languages.
+        """
+        # TODO we need to overweite this
+        # tags = super().get_tags()
```
I'm not sure where's the best place to call this custom class method..

Maybe in `Taxonomy.get_tags`? e.g.,

```
def get_tags(self):
   ObjectTagClass = self.object_tag_class  # might need to catch ImportError..
   if ObjectTagClass and hasattr(ObjectTagClass, 'get_tags'):
       return ObjectTagClass.get_tags(self)

   ...
```

ChrisChV replied:
@pomegranited I think it would be better if the `ObjectTag` function returns a filtered query set, which is used in `Taxonomy.get_tags` . This is so that `Taxonomy.get_tags` is always consistent in the data it returns.

**LanguageObjectTag.get_tags_query_set**

```
class LanguageObjectTag(ClosedSystemObjectTag):
@classmethod
    def get_tags_query_set(cls, taxonomy: Taxonomy) -> models.QuerySet:
        """
        Returns a query set of available languages tags.
        """
        available_langs = cls._get_available_languages()
        return taxonomy.tag_set.filter(external_id__in=available_langs)
```

**Taxonomy.get_tags**

```
def get_tags(self):
   query_set = self.tag_set
   ObjectTagClass = self.object_tag_class  # might need to catch ImportError..
   if ObjectTagClass and hasattr(ObjectTagClass, 'get_tags_query_set'):
       query_set = ObjectTagClass.get_tags_query_set(self)

   # Use query set in operations here
```

ChrisChV committed:
chore: get_tags_query_set added to LanguageObjectTag

ChrisChV committed:
chore: Adding _validate_taxonomy function to all system defined object tags

ChrisChV commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,200 @@
+    - LanguageBlockObjectTag
+
+    On the example, there are ObjectTags for the same Language
+    taxonomy but with different objects.
```
@pomegranited I spoted this case, there can be several `ObjectTags` for the same Taxonomy. This makes `_object_tag_class` not work in this cases. I've solved that here: https://github.com/open-craft/openedx-learning/pull/3/commits/b67ca9a6061fce7fe7f1d61f51e06bfb757edf25

What do you think?

pomegranited replied:
@ChrisChV I'm confused.. can't we create separate Taxonomy instances for each of the System taxonomies, and set a different Taxonomy._object_tag_class for each of them? 

I went back and forth on whether we even needed the ObjectTag registry, because in theory, every new Taxonomy could just specify its ideal ObjectTagClass on creation. But I thought that might add difficulty for the Taxonomy Admins, because they'd have to understand the ObjectTag classes. But since we're doing the System taxonomies as separate hardcoded things, we can include the `_object_tag_class` in the fixtures.

ChrisChV replied:
@pomegranited I saw this in your [edx-platform PR](https://github.com/openedx/edx-platform/pull/32661/files#diff-8a644422ddd022128d13b6ca7c8b672a33d048d1816a80a80491246bdd76f607R64). I saw that depending on the object, one validation or another is necessary (Block or course). In the specific case of Languages, if we create a Taxonomy instance for each Object, each instance would also have to have all the language Tags (I think this also happens in closed taxonomies created by the user)

My idea was to create only one instance of Languages and connect it to the ObjectTags using `system_defined_taxonomy_name` ([like this](https://github.com/open-craft/openedx-learning/pull/3/files/b67ca9a6061fce7fe7f1d61f51e06bfb757edf25#diff-2ecc1274a15be58e16fa617e6e5860e1b2815b6cbeffa200daea9676c07c4bfcR159)) or via the taxonomy id. So when I create `LanguageCourseObjectTag` and `LanguageBlockObjectTag`, both are valid for the Language Taxonomy instance, it would depend on the object, which class to take. 

Is there an advantage to having one instance of the Taxonomy for each object?

pomegranited replied:
Ok @ChrisChV , I think I've addressed this issue with the latest changes.. see https://github.com/openedx/edx-platform/pull/32661#discussion_r1260655868.

ChrisChV committed:
chore: Updating system_defined_taxonomy_id

ChrisChV committed:
refactor: consolidates ObjectTag validation

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+    """
+    System-defined taxonomy IDs
+    """
+    LanguageTaxonomy = 1
```
Ok here is a good reason to add `uuid` as a primary key field for Taxonomy and Tag. I considered that during the architecture discovery but discarded it because I didn't think we needed them.. but here, we do.

Because with UUIDs, we can generate random, unique identifiers that aren't sequential, and so later, once people have added their own taxonomies and we find out we need to add a new system taxonomy, we don't have to worry about primary key clashes.

Could you do that for us here? Best to use [`openedx_learning.lib.fields.immutable_uuid_field`](https://github.com/openedx/openedx-learning/blob/a5b109f58dfe27c63033240c6b9458887014f517/openedx_learning/lib/fields.py#L82-L97).

Also: in the future, if someone wants to add a new system taxonomy, or add new tags to an existing one, how should they do it? Once the migration with the `loaddata` commands is run, it's not going to be re-run, even if the `system_defined.yaml` fixture file changes.

So we should work out the procedure and provide docs for it.

bradenmacdonald replied:
An alternate approach that is simpler is just to make the system-defined IDs negative numbers. The default autoincrementing IDs from user-created taxonomies will always be positive, so there won't be a clash.

pomegranited replied:
I really like that idea, thanks @bradenmacdonald !

ChrisChV replied:
Updated here https://github.com/open-craft/openedx-learning/pull/3/commits/70c34e1bdaac85bdf0f9bbc39967cf13e7a31edf

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+            intance_id = int(self.value)
+        except ValueError:
+            return False
+        return self.tag_class_model.objects.filter(id=intance_id).exists()
```
As noted, we can't rely on the primary key always being called `id`, and we don't know whether it's an integer.

So I think this should be:
```suggestion
        return self.tag_class_model.objects.filter(pk=self.value).exists()
```

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+        if not issubclass(self.tag_class_model, models.Model):
+            return False
+
+        # Verify that the model has 'id' field
+        try:
+            self.tag_class_model._meta.get_field('id')
+        except FieldDoesNotExist:
+            return False
+
```
We can't rely on models having a field named `id`, but we can always use the built-in Django field `pk` instead. cf https://stackoverflow.com/a/2165880

```suggestion
```

ChrisChV replied:
Noted, thanks!

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
```
If you switch around the class inheritance for OpenSystemObjectTag (and all the other tags like it), then the code for these subclasses can be even simpler.

Also:
* I'm using `taxonomy.uuid` here instead of `id` too, see [previous comment](https://github.com/open-craft/openedx-learning/pull/3/files#diff-2ecc1274a15be58e16fa617e6e5860e1b2815b6cbeffa200daea9676c07c4bfcR25).
* Moved the `system_defined_taxonomy_id` class constant to an "unimplemented" instance property method instead, so that subclasses are forced to implement it, or face a compile-time error.

```suggestion
class SystemDefinedObjectTagMixin:
    """
    Mix-in for ObjectTags used on all system defined taxonomies

    `system_defined_taxonomy_uuid` is used to connect the
    ObjectTag with the system defined taxonomy.
    This is because there can be several ObjectTags
    for the same Taxonomy, ex:

    - LanguageCourseObjectTag
    - LanguageBlockObjectTag

    On the example, there are ObjectTags for the same Language
    taxonomy but with different object types.

    Using this approach makes the connection between the ObjectTag
    and system defined taxonomy hardcoded and can't be changed.
    """

    @property
    def system_defined_taxonomy_uuid(self) -> uuid4:
        """
        Subclasses must implement this method to return the uuid
        of the system taxonomy used by this ObjectTag subclass.
        """
        raise NotImplementedError

    def _check_system_taxonomy(self) -> bool:
        """
        Validates if the taxonomy is system-defined and matches
        with the UUID stored in the object tag class.
        """
        return (
            self.taxonomy_id and
            self.taxonomy.system_defined and
            self.taxonomy.uuid == self.system_defined_taxonomy_uuid
         )

    def _check_taxonomy(self) -> bool:
        """
        Validates the system-defined taxonomy.
        """
        return (
            super()._check_taxonomy() and
            self._check_system_taxonomy()
        )


class OpenSystemObjectTag(SystemDefinedObjectTagMixin, OpenObjectTag):
    """
    Free-text object tag used on system-defined taxonomies
    """

class ClosedSystemObjectTag(SystemDefinedObjectTagMixin, ClosedObjectTag):
    """
    Object tags linked to a closed system-taxonomy
    """
```

ChrisChV replied:
[Updated here](https://github.com/open-craft/openedx-learning/pull/3/files#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R691)

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+            return False
+    
+        if not self.tag_class_model:
+            return False
```
Similarly here, if we use an "unimplemented" instance property method instead of a class constant, subclasses will be forced to implement it, or face a compile-time error.

Also, because class-level inheritance for python classes isn't so great, using an instance property allows sub-subclasses to override this property as well (if anyone ever wants to do that).

```suggestion

    @property
    def tag_class_model(self) -> Type:
        """
        Subclasses must implement this method to return the Django.model
        class referenced by these object tags.
        """
        raise NotImplementedError

    def _check_taxonomy(self):
        """
        Validates if has an associated Django model that has an Id
        """
        if not super()._check_taxonomy():
            return False
```

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+
+        # Verify that is a Django model
+        if not issubclass(self.tag_class_model, models.Model):
+            return False
```
This should raise an exception instead of just silently failing, to make it easier for developers to figure out what's wrong with their ModelObjectTag subclasses..

I'm not sure if this is the best place to raise this exception though, so am open to other ideas.

```suggestion
        # Verify that is a Django model
        assert issubclass(self.tag_class_model, models.Model)
```

ChrisChV replied:
[Updated here](https://github.com/open-craft/openedx-learning/pull/3/files#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R687)

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+        if not self._check_instance():
+            return False
+        
+        return True
```
I like the shorthand implementation you used above better (I should [do that too](https://github.com/openedx/edx-platform/pull/32661#pullrequestreview-1527743164)):
```suggestion
        return (
            super()._check_tag() and
            self._check_instance()
        )
```

ChrisChV replied:
Updated in all checks

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+        if not self.tag or not self.tag.external_id in available_langs:
+            return False
+        
+        return True
```
Could even do this shorter as well, if you think it's legible enough:
```suggestion
        return (
            super()._check_tag() and 
            # Must be linked to a tag with an available language
            self.tag and self.tag.external_id in self._get_available_languages()
        )
```

pomegranited commented on tests/openedx_tagging/core/tagging/test_system_defined.py:
```diff
@@ -0,0 +1,236 @@
+
+    class Meta:
+        proxy = True
+        managed = False
```
Cool.. I didn't know you could do this in tests!

ChrisChV replied:
Thanks :smile:

pomegranited commented on tests/openedx_tagging/core/tagging/test_system_defined.py:
```diff
@@ -0,0 +1,236 @@
+    """
+    Test for generic system defined object tags
+    """
+    def test_system_defined_is_valid(self):
+        # Valid
+        assert TestUserObjectTag()._check_system_taxonomy(taxonomy=self.user_system_taxonomy)
+
+        # Null taxonomy
+        assert not UserObjectTag()._check_system_taxonomy()
+
+        # Not system defined taxonomy
+        assert not UserObjectTag()._check_system_taxonomy(taxonomy=self.taxonomy)
+
+        # Not connected with the taxonomy
+        assert not UserObjectTag()._check_system_taxonomy(taxonomy=self.user_system_taxonomy)
```
Can use ddt, avoid tying your tests to internal methods, and deal with the [changes I suggested above](https://github.com/open-craft/openedx-learning/pull/3/files#diff-2ecc1274a15be58e16fa617e6e5860e1b2815b6cbeffa200daea9676c07c4bfcR88) by testing this way instead:

```suggestion
@ddt.data(
    "None",
    "taxonomy",
    "user_system_taxonomy",
)
def test_system_defined_is_valid(self, taxonomy_attr):
        taxonomy = getattr(self, taxonomy_attr, None)
        object_tag =  TestUserObjectTag(taxonomy=taxonomy)
        assert not object_tag.is_valid(check_taxonomy=True, check_tag=False, check_object=False)
```

ChrisChV replied:
All test has been updated to use ddt

pomegranited commented on tests/openedx_tagging/core/tagging/test_system_defined.py:
```diff
@@ -0,0 +1,236 @@
+            value=value,
+        )
+
+        result = object_tag.is_valid(
+            check_taxonomy=True,
+            check_object=True,
+            check_tag=True,
+        )
+
+        self.assertEqual(result, assert_value)
```
It's ok that all of these are explicitly laid out, but you can just say this if you want to:
```suggestion
        assert object_tag.is_valid() == assert_value
```

ChrisChV replied:
All test has been updated with this

@pomegranited In your experience, is there a difference in using one approach or another? For my tests I have always used `self.assertEqual()`

pomegranited replied:
@ChrisChV There's no functional difference.. I just think using plain python wherever possible looks cleaner. But I'm not that fussy about it -- we're using TestClass as a base, so all of its `.assert*` methods are available regardless.

pomegranited commented on tests/openedx_tagging/core/tagging/test_system_defined.py:
```diff
@@ -0,0 +1,236 @@
+            'value': 'value',
+        }
+        result = object_tag(**args).is_valid(check_object=False, check_tag=False, check_taxonomy=True)
+        self.assertEqual(result, assert_value)
```
nit: can do this with a plain assert check:
```suggestion
        assert result == assert_value
```

ChrisChV replied:
All test has been updated with this

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+        Returns a query set of available languages tags.
+        """
+        available_langs = cls._get_available_languages()
+        return taxonomy.tag_set.filter(external_id__in=available_langs)
```
If we make this a classmethod on ObjectTag too, then we can call it in `Taxonomy.get_tags` without have to check whether or not the method exists on the subclass.

So in ObjectTag, add:

```python
    @classmethod
    def get_tags_query_set(cls, taxonomy: Taxonomy) -> models.QuerySet:
        """
        Returns a base query set of the tags available on the given taxonomy.
        
        Subclasses should override this method if they need to provide their own custom tag sets.
        """
        return taxonomy.tag_set
```

And so Taxonomy's `get_tags` method can be updated as you recommended:

```python
    def get_tags(self) -> List[Tag]:
        """
        Returns a list of all Tags in the current taxonomy, from the root(s) down to TAXONOMY_MAX_DEPTH tags, in tree order.

        Annotates each returned Tag with its ``depth`` in the tree (starting at 0).

        Performance note: may perform as many as TAXONOMY_MAX_DEPTH select queries.
        """
        tags = []
        if self.allow_free_text:
            return tags

        ObjectTagClass = self.object_tag_class or ObjectTag
        tag_set = ObjectTagClass.get_tags_query_set(self).prefetch_related("parent")
        parents = None
        for depth in range(TAXONOMY_MAX_DEPTH):
            filtered_tags = tag_set
```

pomegranited replied:
Edit: Since we can now subclass Taxonomies, you can just make your custom Taxonomy override get_tags, to set the base queryset.

ChrisChV replied:
Updated here https://github.com/open-craft/openedx-learning/pull/3/commits/62501cddbab4ed5b2360be061acbeb9b56436fff

pomegranited commented on tests/openedx_tagging/core/tagging/test_system_defined.py:
```diff
@@ -0,0 +1,236 @@
+    def test_get_tags_query_set(self):
+        tags = LanguageObjectTag.get_tags_query_set(self.taxonomy)
+        for tag in tags:
+            self.assertIn(tag.value, self.expected_langs_values)
```
Could you also add some tests [like these](https://github.com/openedx/openedx-learning/pull/62/files#diff-6cda6f74e9f6bf053923d015688ea76e50cfec92ef244df48aaa3942288fcbb5R348-R385) to ensure that the `tagging_api.cast_object_tag` method casts the system language tags to your `LanguageObjectTag` class as expected?

Since everything hinges on us using the correct ObjectTag subclass, we need to make sure this remains true.

pomegranited replied:
This comment is obsolete now.. closing.

pomegranited reviewed (changes requested):
Hi @ChrisChV, this is really interesting.. I couldn't see how the system taxonomies would fit together until I read your code! Very cool.

I've requested some changes, but:  Feel free to wait until we've got approval at least in principle on https://github.com/openedx/openedx-learning/pull/62 before taking time to apply them, just in case that approach gets changed again.

Also, feel free to push back on anything that seems unreasonable?

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+    class Meta:
+        proxy = True
+
+    tag_class_model = get_user_model()
```
I really like how you used `external_id` to store the language code for the language object tag , and then used the full language name as the value. So when we actually fetch an ObjectTag for display, we have a nice human-readable value to show.

I'd like to support something like that here with model tags. So that when we have a tag that references a User or Organization, we can show the User/Organization full name instead of the ID. Plus, that name will always be up-to-date, because we're fetching it from the source table.

So I'm wondering, can we do something like this:

```python
class ObjectTag:
    @property
    def value_pretty(self):
        """
        Returns a human-friendly version of the tag value stored here.
        
        Subclasses should override this if they don't store human-readable values.
        """
        return self.value
```

And then we can have here:
```python
class ModelObjectTag(...):
    @property
    def tag_class_model_pretty_field(self):
        """
        Name of the field to use on the tag_class_model instance to which stores the "pretty value" of the tag.
        """
        raise NotImplementedError

    @property
    def value_pretty(self):
        """
        Returns a human-friendly version of the tag value stored here.
        
        Subclasses should override this if they don't store human-readable values.
        """
        value = self.value
        try:
            instance = self.tag_class_model.objects.get(pk=value)
            value = getattr(instance, self.tag_class_model_pretty_field, value)
        except self.tag_class_model.DoesNotExist:
            # Means the tag is invalid, but that's ok here
            pass
            
        return value
...

class UserObjectTag(ModelObjectTag):
    @property
    def tag_class_model_pretty_field(self):
        """
        Name of the field to use on the tag_class_model instance to which stores the "pretty value" of the tag.
        """
        return "full_name"
```

ChrisChV replied:
Updated to use this approach https://github.com/open-craft/openedx-learning/pull/3#discussion_r1267622247

bradenmacdonald commented on openedx_tagging/core/tagging/fixtures/language_taxonomy.yaml:
```diff
@@ -0,0 +1,1288 @@
+  fields:
+    taxonomy: 1
+    parent: null
+    value: German
```
Please make sure we're thinking about how to translate these before displaying them in the UI. For a German user, this language tag needs to appear as "Deutsch", not "German".

We should probably use a similar mechanism for users, storing their user ID as the tag value but displaying their username or their "full name (username)" in the UI.

You don't have to implement this in this PR (it's a bad idea to, actually) - just think about how it will work in the future.

pomegranited replied:
> We should probably use a similar mechanism for users, storing their user ID as the tag value but displaying their username or their "full name (username)" in the UI.

What do you think of [this idea](https://github.com/open-craft/openedx-learning/pull/3#discussion_r1262170302)?

I think it could be adapted for translations too..

bradenmacdonald replied:
I think something like that would make sense, yes. Just not clear on how it would be adapted to the current approach. But let's not worry too much about that just yet; I don't want to muddy the waters any further here :)

pomegranited commented on openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py:
```diff
@@ -0,0 +1,210 @@
+        )
+
+
+class ModelObjectTag(OpenSystemObjectTag):
```
@ChrisChV Since we've gone back to subclassing Taxonomy again, I wonder if we could create a ModelTaxonomy class, which can:

* Create Tag objects on the fly when new ObjectTags are added
   * set external_id=model.pk
   * set value = value we want to show to the world, e.g. User.username or Organization.name
   * "Resync" could somehow re-fetch the related object and update this value? Or maybe we need another method.
* Can the ModelTaxonomy can be "closed" even though the Tag values don't exist (yet)?
* Can we return "autocomplete_tags" based on the values available in the linked model?

ChrisChV replied:
> Create Tag objects on the fly when new ObjectTags are added

This has been one of the [rejected options](https://github.com/openedx/openedx-learning/blob/main/docs/decisions/0012-system-taxonomy-creation.rst#tags-created-by-auto-generated-from-the-codebase) for creating Tags. But now that I see it, it's easier than it seemed and has many more advantages. Yes, I will build this type of taxonomy

ChrisChV replied:
@pomegranited I updated the ADR https://github.com/open-craft/openedx-learning/pull/3/commits/e031a5e735ea247e2709d16440ca45d71c767632

pomegranited committed:
refactor: use negative numbers as primary keys for system taxonomies and tags

pomegranited committed:
style: linting and formatting

pomegranited committed:
refactor: use ObjectTag subclasses where possible

* Adds Taxonomy.object_tag_class property, which can be overridden by
  Taxonomy subclasses to provide their own ObjectTag subclasses.
* Replaces the Taxonomy._set_tag method with an ObjectTag.tag_ref
  property setter, so that ObjectTag subclasses can dynamically create
  Tags as needed during tag_object()
* Moves the tag_class_model and tag_class_value properties to a new
  ModelObjectTag abstract class, so the ModelSystemDefinedTaxonomy can
  be simpler.
* Updates tests for coverage

pomegranited committed:
refactor: LanguageTaxonomy overrides get_tags

removing the need for _get_tags_query_set

ChrisChV committed:
feat: System defined taxonomies subclasses

pomegranited committed:
Allow custom Taxonomy, ObjectTag subclasses to customize tagging behavior (#62)

Adds support for custom Taxonomy subclasses to be stored against a Taxonomy, to be used by the python API when instantiating and returning Taxonomies.

Also adds minimal support for ObjectTag subclasses. However, these are not stored against the ObjectTag instances; they can be instantiated by the Taxonomy subclasses if and when needed.

Related:
* docs: updates decisions to reflect this change
* feat: adds api.get_taxonomy, which returns a Taxonomy cast to its subclass, when set
* refactor: adds _check_taxonomy, _check_tag, and _check_object methods to the Taxonomy class, which can be overridden by subclasses when validating ObjectTags

Added to support system-defined Taxonomies:
* feat: adds un-editable Taxonomy.system_defined field so that system taxonomies can store this field and ensure no one edits them. 
* feat: adds Taxonomy.visible_to_authors, which is needed for fully automated tagging.

Cleanup changes:
* fix: updates Tag model to cascade delete if the Taxonomy or parent Tag is deleted.
* style: adds missing type annotations to rules and python API

pomegranited commented on openedx_tagging/core/tagging/fixtures/system_defined.yaml:
```diff
@@ -0,0 +1,11 @@
+- model: oel_tagging.taxonomy
+  pk: 1
```
I liked @bradenmacdonald 's idea to use negative IDs for system-defined taxonomies and their tags, to avoid clashing with user-created taxonomies and tags.

See https://github.com/open-craft/openedx-learning/pull/6/commits/af8910a5979d6006189bef6519f8023dd038c79a for suggested changes, but I also think it's worth amending decision [12. System-defined Taxonomy & Tags creation](https://github.com/openedx/openedx-learning/blob/main/docs/decisions/0012-system-taxonomy-creation.rst) to add this note in.

ChrisChV replied:
Updated here https://github.com/open-craft/openedx-learning/pull/3/commits/e031a5e735ea247e2709d16440ca45d71c767632

pomegranited commented on openedx_tagging/core/tagging/models.py:
```diff
@@ -627,3 +649,240 @@ def copy(self, object_tag: "ObjectTag") -> "ObjectTag":
+    Model based system taxonomy abstract class.
+    
+    This type of taxonomies has an associated Django model in `tag_class_model()`.
+    Are designed to create a Tag when an ObjectTags with a new Tag. 
```
nit:
```suggestion
    This type of taxonomy has an associated Django model in `tag_class_model()`.
    They are designed to create Tags when required for new ObjectTags, to maintain their status as "closed" taxonomies.
```

ChrisChV replied:
Updated here https://github.com/open-craft/openedx-learning/pull/3/commits/55ea40ae54637eaa423ef7dc6f81f05604af39af

pomegranited commented on openedx_tagging/core/tagging/models.py:
```diff
@@ -627,3 +649,240 @@ def copy(self, object_tag: "ObjectTag") -> "ObjectTag":
+    The Tags are representations of the instances of the associated model.
+
+    On Tag.external_id stores an identifier from the instance (`pk` as default)
+    and on Tag.value stores an human readable representation of the instance
```
nit:
```suggestion
    Tag.external_id stores an identifier from the instance (`pk` as default)
    and Tag.value stores a human readable representation of the instance
```

ChrisChV replied:
Updated here https://github.com/open-craft/openedx-learning/pull/3/commits/55ea40ae54637eaa423ef7dc6f81f05604af39af

pomegranited commented on openedx_tagging/core/tagging/models.py:
```diff
@@ -627,3 +649,240 @@ def copy(self, object_tag: "ObjectTag") -> "ObjectTag":
+        except self.tag_class_model.DoesNotExist:
+            return None
+    
+    def _resync_tag(self, tag: Tag) -> Tag:
```
I think it's cleaner to define a `ModelObjectTag` subclass of ObjectTag to handle most of this functionality, cf https://github.com/open-craft/openedx-learning/pull/6/commits/9fd8854c293cc9b506c2a55181c5aab9f6b2b80d

ChrisChV replied:
Added to the branch

pomegranited commented on openedx_tagging/core/tagging/models.py:
```diff
@@ -627,3 +649,240 @@ def copy(self, object_tag: "ObjectTag") -> "ObjectTag":
+        self,
+        tag_ref,
+        object_tag: ObjectTag
+    ) -> ObjectTag:
```
I created a [property setter for `ObjectTag.tag_ref`](https://github.com/open-craft/openedx-learning/pull/6/commits/9fd8854c293cc9b506c2a55181c5aab9f6b2b80d#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R545) which handles the dynamic Tag creation like you've done here.

It's [cleaner to call](https://github.com/open-craft/openedx-learning/pull/6/commits/9fd8854c293cc9b506c2a55181c5aab9f6b2b80d#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R404), and I think makes more sense from an encapsulation perspective.

ChrisChV replied:
Added to the branch

pomegranited commented on openedx_tagging/core/tagging/models.py:
```diff
@@ -627,3 +649,240 @@ def copy(self, object_tag: "ObjectTag") -> "ObjectTag":
+        """
+        Returns the username as tag value
+        """
+        return instance.get_username()
```
Moved this functionality into [UserModelObjectTag](https://github.com/open-craft/openedx-learning/pull/6/commits/9fd8854c293cc9b506c2a55181c5aab9f6b2b80d#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R811), which makes [UserSystemDefinedTaxonomy simpler](https://github.com/open-craft/openedx-learning/pull/6/commits/9fd8854c293cc9b506c2a55181c5aab9f6b2b80d#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R836).

ChrisChV replied:
Added to the branch

pomegranited commented on openedx_tagging/core/tagging/models.py:
```diff
@@ -627,3 +649,240 @@ def copy(self, object_tag: "ObjectTag") -> "ObjectTag":
+        Returns a query set of available languages tags.
+        """
+        available_langs = self._get_available_languages()
+        return self.tag_set.filter(external_id__in=available_langs)
```
Since LanguageTaxonomy is a subclass of Taxonomy, I think it's cleaner to [do this instead](https://github.com/open-craft/openedx-learning/pull/6/commits/bc2c31788d67b842cd26121a9dc67e17633ef18c).

ChrisChV replied:
Added to the branch

pomegranited commented on tests/openedx_tagging/core/tagging/test_models.py:
```diff
@@ -402,3 +466,189 @@ def test_tag_object_invalid_tag(self):
+
+
+@ddt.ddt
+class TestSystemDefinedTaxonomy(TestTagTaxonomyMixin, TestCase):
```
If you have time, could you a test for a simple, fully dynamic SystemTaxonomy as @bradenmacdonald [suggested here](https://github.com/openedx/openedx-learning/pull/62#discussion_r1267116903)?

pomegranited reviewed (changes requested):
Hi @ChrisChV , I started suggesting things, but then decided it would be clearer to put the changes into a PR: https://github.com/open-craft/openedx-learning/pull/6

What do you think about that approach instead?

I'm also thinking we could rearrange the model code a bit to shorten the (now quite long) `models.py`:

* move the base models (Tag, Taxonomy, ObjectTag) to `openedx_tagging/core/tagging/models/base.py`
* move the system-defined models to `openedx_tagging/core/tagging/models/system_defined.py`
* import the models we want people to use outside this library (Tag, Taxonomy, ObjectTag, LanguageTaxonomy, UserModelTaxonomy, etc.) in `openedx_tagging/core/tagging/models/__init__.py`

If you agree, feel free to split out a `test_system_defined_models.py` as well.

pomegranited committed:
fix: removes COMPRESSED row format from openedx_learning core (#66)

AWS Aurora does not support COMPRESSED in its MySQL implementation:

https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Migrating.RDSMySQL.Import.html

ChrisChV committed:
docs: System taxonomy creation doc updated with Dynamic tags approach

ChrisChV committed:
style: Updating comments

ChrisChV commented:
> I'm also thinking we could rearrange the model code a bit to shorten the (now quite long) models.py:

@pomegranited  Yes, I agree with that. Thanks. I will add it

ChrisChV committed:
style: Separating the models into base and system defined

ChrisChV committed:
fix: Update language fixture to negative pk

ChrisChV committed:
feat: Updates on Taxonomy and Tag admins

- Taxonomy.system_defined updated to be a hardcoded property.
- TaxonomyAdmin updated to avoid edit and delete system taxonomies.
- TagAdmin updated to avoid edit and delete tags from system taxonomies.
- TagAmdin updated to avoid create Tags on system taxonomies.

ChrisChV committed:
feat: Instance validations on ModelSystemDefinedTaxonomy

pomegranited commented:
Closed in favor of https://github.com/openedx/openedx-learning/pull/67
